src/msg/async/dpdk/DPDK.cc

```diff
@@ -644,6 +669,90 @@ DPDKQueuePair::DPDKQueuePair(CephContext *c, EventCenter *cen, DPDKDevice* dev,
     device_stat_time_fd = center->create_time_event(1000*1000, new C_handle_dev_stats(this));
 }
 
+void DPDKDevice::nic_stats_dump(Formatter *f)
+{
+  static uint64_t prev_pkts_rx[RTE_MAX_ETHPORTS];
+  static uint64_t prev_pkts_tx[RTE_MAX_ETHPORTS];
+  static uint64_t prev_cycles[RTE_MAX_ETHPORTS];
+  uint64_t diff_pkts_rx, diff_pkts_tx, diff_cycles;
+  uint64_t mpps_rx, mpps_tx;
+  struct rte_eth_stats stats;
+  size_t tx_fragments = 0;
+  size_t rx_fragments = 0;
+  size_t tx_free_cnt = 0;
+  size_t rx_free_cnt = 0;
+
+  for (auto &qp: _queues) {
+    tx_fragments += qp->perf_logger->get(l_dpdk_qp_tx_fragments);
+    rx_fragments += qp->perf_logger->get(l_dpdk_qp_rx_fragments);
+    tx_free_cnt += qp->_tx_buf_factory.ring_size();
+    rx_free_cnt += rte_mempool_avail_count(qp->_pktmbuf_pool_rx);
+  }
+  rte_eth_stats_get(_port_idx, &stats);
+  f->open_object_section("stats");
+  f->dump_unsigned("RX-packets", stats.ipackets);
+  f->dump_unsigned("DPDK RX-packets", rx_fragments);
+  f->dump_unsigned("RX-bytes", stats.ibytes);
+  f->dump_unsigned("RX-missed", stats.imissed);
+  f->dump_unsigned("RX-errors", stats.ierrors);
+
+  f->dump_unsigned("TX-packets", stats.opackets);
+  f->dump_unsigned("DPDK TX-packets", tx_fragments);
+  f->dump_unsigned("TX-bytes", stats.obytes);
+  f->dump_unsigned("TX-errors", stats.oerrors);
+
+  f->dump_unsigned("RX-nombuf", stats.rx_nombuf);
+  f->dump_unsigned("RX-avail mbufs", rx_free_cnt);
+  f->dump_unsigned("TX-avail mbufs", tx_free_cnt);
+  diff_cycles = prev_cycles[_port_idx];
+  prev_cycles[_port_idx] = rte_rdtsc();
+  if (diff_cycles > 0)
+    diff_cycles = prev_cycles[_port_idx] - diff_cycles;
+
+  diff_pkts_rx = (stats.ipackets > prev_pkts_rx[_port_idx]) ?
+	         (stats.ipackets - prev_pkts_rx[_port_idx]) : 0;
+  diff_pkts_tx = (stats.opackets > prev_pkts_tx[_port_idx]) ?
+	         (stats.opackets - prev_pkts_tx[_port_idx]) : 0;
+  prev_pkts_rx[_port_idx] = stats.ipackets;
+  prev_pkts_tx[_port_idx] = stats.opackets;
+  mpps_rx = diff_cycles > 0 ? diff_pkts_rx * rte_get_tsc_hz()/diff_cycles : 0;
```

Comment:
could you please define variable at where it is used for the first time, for better readability?